Fix CMake FindPDAL for PDAL version 2.6+

nilason opened the pull request:
## Description

The newly released PDAL 2.6 removed the 'pdal_util' library as a separate library, it is now part of the 'pdalcpp' library (see https://github.com/PDAL/PDAL/releases/tag/2.6.0). The CMake FindPDAL thus fails to locate this now non-existing library. This PR addresses this issue and should be backported to 3.28 and 3.32.


Note, I added a style formatting commit for the FindPDAL.cmake file to this PR  with the intent to modernise the code and improve readability. If you find that particular change is not appropriate, I'll remove that part.



<!--
  BEFORE HITTING SUBMIT -- Please BUILD AND TEST your changes thoroughly. This is YOUR responsibility! Do NOT rely on the QGIS code maintainers to do this for you!!

  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

  Also mention if you think this PR needs to be backported. And list relevant or fixed issues.

------------------------

  Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by checking the following list.
  Feel free to ask in a comment if you have troubles with any of them.

  - Commit messages are descriptive and explain the rationale for changes.

  - Commits which fix bugs include `Fixes #11111` at the bottom of the commit message. If this is your first pull request and you forgot to do this, write the same statement into this text field with the pull request description.

  - New unit tests have been added for relevant changes

  - You have run the `scripts/prepare_commit.sh` script (https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit.
    If you didn't do this, you can also run `./scripts/astyle_all.sh` from your source folder.

  - You have read the QGIS Coding Standards (https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
-->

m-kuhn commented:
Don't newer PDAL version already ship a pdalconfig.cmake file which we could rely on and only use the old code as fallback?
Cmp. what has been done for Proj etc https://github.com/qgis/QGIS/commit/789e92b05436f3d1f6d3ca8950a74265892090d3#diff-2fa2a6c4b1439910cf5be5f527f911fd4c752870c0b999306dbdcd61df63211bL21

nilason committed:
cmake: update to find PDAL 2.6+

With PDAL 2.6 the 'pdal_util' library is removed as a separate library
and hence part of the 'pdalcpp' library.

nilason commented:
> Don't newer PDAL version already ship a pdalconfig.cmake file which we could rely on and only use the old code as fallback? Cmp. what has been done for Proj etc [789e92b#diff-2fa2a6c4b1439910cf5be5f527f911fd4c752870c0b999306dbdcd61df63211bL21](https://github.com/qgis/QGIS/commit/789e92b05436f3d1f6d3ca8950a74265892090d3#diff-2fa2a6c4b1439910cf5be5f527f911fd4c752870c0b999306dbdcd61df63211bL21)

Actually, PDAL did almost always come with cmake config-files. Thanks for the Proj solution link. I have now updated this PR accordingly. Tested with PDAL 2.5.6 and 2.6.0.

I wonder if  the:
```cmake
if (WITH_PDAL AND MSVC)
  add_definitions(-DWIN32_LEAN_AND_MEAN)
endif()
```
parts cannot be moved to FindPDAL.cmake? It is now repeated in three files.

m-kuhn commented on src/app/CMakeLists.txt:
```diff
@@ -482,11 +482,8 @@ if(HAVE_OPENCL)
-    add_definitions(-DWIN32_LEAN_AND_MEAN)
-  endif()
+if (WITH_PDAL AND MSVC)
+  add_definitions(-DWIN32_LEAN_AND_MEAN)
```
I guess this should go to FindPDAL as well with 
[`target_compile_definitions`](https://cmake.org/cmake/help/v3.2/command/target_compile_definitions.html)

nilason replied:
That would probably not work as the target *is* imported:
> The named <target> must have been created by a command such as [add_executable()](https://cmake.org/cmake/help/v3.2/command/add_executable.html#command:add_executable) or [add_library()](https://cmake.org/cmake/help/v3.2/command/add_library.html#command:add_library) and must not be an [Imported Target](https://cmake.org/cmake/help/v3.2/manual/cmake-buildsystem.7.html#imported-targets).

nilason replied:
It might work for the fallback code. Wouldn't that part be included with the config-files for Windows?

m-kuhn replied:
Yes, that should be part of the fallback code or defined by the config files in the Windows case

nilason replied:
> ...  or defined by the config files in the Windows case

Well, is it?

And is `WIN32_LEAN_AND_MEAN` needed at all nowadays (just looked at this 14-year-old post https://devblogs.microsoft.com/oldnewthing/20091130-00/?p=15863)?
Maybe @jef-n can help us here?

jef-n replied:
it is: https://cdash.orfeo-toolbox.org/build/18183

nilason replied:
> That would probably not work as the target _is_ imported:
> 
> > The named `<target>` must have been created by a command such as [add_executable()](https://cmake.org/cmake/help/v3.2/command/add_executable.html#command:add_executable) or [add_library()](https://cmake.org/cmake/help/v3.2/command/add_library.html#command:add_library) and must not be an [Imported Target](https://cmake.org/cmake/help/v3.2/manual/cmake-buildsystem.7.html#imported-targets).

That quote was from a very old cmake version docs. If I'm not mistaken, minimal CMake version required for QGIS is 3.12. In that case `target_compile_definitions` seems to work.

nilason replied:
I added a commit with all windows definitions of WIN32_LEAN_AND_MEAN moved to FindPDAL (using target_compile_definitions).

m-kuhn commented on cmake/FindPDAL.cmake:
```diff
@@ -7,8 +7,7 @@
-# and following variables are set:
-#    PDAL_INCLUDE_DIR
+# and adds the following target(s):
 #    PDAL_LIBRARIES
```
```suggestion
#    - pdalcpp
#    - pdal_util
```

nilason replied:
The PDAL config-files defines `PDAL_LIBRARIES`, which may be either a list `pdalcpp;pdal_util` (<2.6) or only `pdalcpp` (2.6+).  So PDAL_LIBRARIES should stay, but I may add a clarifying comment.

m-kuhn replied:
Good point, let's keep as it is then 👍

nilason committed:
set -DWIN32_LEAN_AND_MEAN in FindPDAL.cmake

agiudiceandrea commented:
It looks like this PR should have been backported to the release-3_28 brach: please see https://github.com/qgis/QGIS/issues/55431.

imincik commented:
Is it possible to backport this change to `release-3_28` so we can build QGIS LTR with PDAL 2.6 ? Thanks
